### PR TITLE
fix: Specifying reason in expect(clippy::needless_return) no longer triggers false positive

### LINKED ISSUE #13366: Specifying reason in `expect(clippy::needless_return)` triggers false positive "this lint expectation is unfulfilled"
### Summary

It seems the `reason` option has an effect on whether the lint warns about an unfulfilled expectation.

### Lint Name

clippy::needless_return

### Reproducer

I tried this code:

```rust
fn a(x: Option<u8>) -> Option<u8> {
    match x {
        Some(_) => None,
        None => {
            #[expect(clippy::needless_return, reason = "Use early return for errors.")]
            return None;
        }
    }
}

fn b(x: Option<u8>) -> Option<u8> {
    match x {
        Some(_) => None,
        None => {
            #[expect(clippy::needless_return)]
            return None;
        }
    }
}
```

I saw this happen:
The warning about unfulfilled expectations is printed for function `a`, but not for `b`:

```
  |
5 |             #[expect(clippy::needless_return, reason = "Use early return for errors.")]
  |                      ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: Use early return for errors.
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

I expected to see this happen:
No warning should be printed for either function.

### Version

```text
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: x86_64-unknown-linux-gnu
release: 1.81.0
LLVM version: 18.1.7
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by vHugoObject**
fixes #13366 
changelog: none

### CODE REVIEW

File: clippy_lints/src/returns.rs [hunk 1]
```diff
@@ -418,19 +418,29 @@ fn check_final_expr<'tcx>(
...
-                        && tool.ident.name == sym::clippy
-                        && matches!(
-                            lint_name.ident.name.as_str(),
-                            "needless_return" | "style" | "all" | "warnings"
-                        )
-                    {
-                        // This is an expectation of the `needless_return` lint
-                    } else {
+                    if !matches!(Level::from_attr(attr), Some(Level::Expect(_))) {
+                        return;
```
Comment (alex-semenyuk) [inline clippy_lints/src/returns.rs:422]: Imho do not think this approach with return for each conditions better then just list of them with &&

**Comment by vHugoObject** [comment]
The false positive was being triggered by a condition that was in the
middle of the list. Early returns made it easier to debug.

File: clippy_lints/src/returns.rs (hunk 1 above)
Comment (vHugoObject) [reply clippy_lints/src/returns.rs:422]: The false positive was being triggered by a condition that was in the
middle of the list. Early returns made it easier to debug.

Comment (vHugoObject) [reply clippy_lints/src/returns.rs]: I reverted back to a list of conditions with &&

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Thanks!

@bors r+

**Pull request merged**